litestar/channels/memory.py

```diff
@@ -0,0 +1,77 @@
+from __future__ import annotations
+
+import asyncio
+import contextlib
+from asyncio import Queue
+from collections import defaultdict, deque
+from typing import Any, AsyncGenerator, Iterable
+
+from litestar.channels.base import ChannelsBackend
+
+
+class MemoryChannelsBackend(ChannelsBackend):
+    """An in-memory channels backend"""
+
+    def __init__(self, history: int = 0) -> None:
+        self._max_history_length = history
+        self._channels: set[str] = set()
+        self._queue: Queue[tuple[str, bytes]] | None = None
+        self._history: defaultdict[str, deque[bytes]] = defaultdict(lambda: deque(maxlen=self._max_history_length))
+
+    async def on_startup(self) -> None:
+        self._queue = Queue()
+
+    async def on_shutdown(self) -> None:
+        self._queue = None
+
+    async def publish(self, data: bytes, channels: Iterable[str]) -> None:
+        """Publish ``data`` to ``channels``. If a channel has not yet been subscribed to,
+        this will be a no-op.
+
+        Args:
+            data: Data to publish
+            channels: Channels to publish to
+
+        Returns:
+            None
+
+        Raises:
+            RuntimeError: If ``on_startup`` has not been called yet
+        """
+        if not self._queue:
+            raise RuntimeError("Backend not yet initialized. Did you forget to call on_startup?")
+
+        for channel in channels:
+            if channel not in self._channels:
+                continue
+            self._queue.put_nowait((channel, data))
+        if self._max_history_length:
+            for channel in channels:
+                self._history[channel].append(data)
+
+    async def subscribe(self, channels: Iterable[str]) -> None:
+        """Subscribe to ``channels``, and enable publishing to them"""
+        self._channels.update(channels)
+
+    async def unsubscribe(self, channels: Iterable[str]) -> None:
+        """Unsubscribe from ``channels``"""
+        self._channels = self._channels - (set(channels))
+        for channel in channels:
+            with contextlib.suppress(KeyError):
```

Comment:
You can move this above the for loop

Reply:
Actually this should go completely. Ruff forces this usage, but I don't like introducing a context stack for just a single operation. Especially since we have zero-cost abstractions in newer Python versions, this seems like an antipattern.